### PR TITLE
Add "IsWrittenTo" from Roslyn

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Pre-Condition for "ArgumentDescriptor"

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs:366]: To be consistent, I moved some of these pre-existing extension methods to the "Roslyn" file.

File: analyzers/src/SonarAnalyzer.Common/Helpers/CodeCopiedFromAttribute.cs
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/CodeCopiedFromAttribute.cs:21]: I suggest to find a better name and place for this attribute.

I don't have super ideas, but "Helpers" is a code smell and having a folder that contains both "meta-programming" and "functional helpers" is not great

File: analyzers/src/SonarAnalyzer.Common/Helpers/CodeCopiedFromAttribute.cs
Comment (antonioaversa) [inline analyzers/src/SonarAnalyzer.Common/Helpers/CodeCopiedFromAttribute.cs:29]: Given that `AttributeTargets` is a `Flags` enum, what about not restricting it, and use `AttributeTargets.All`?
Less precise maybe, but doesn't require to list all possible targets of copy.

File: analyzers/src/SonarAnalyzer.Common/Helpers/CodeCopiedFromAttribute.cs
Comment (antonioaversa) [inline analyzers/src/SonarAnalyzer.Common/Helpers/CodeCopiedFromAttribute.cs:30]: Could you give an example when you would like to be able to add the attribute multiple times for the same member? I think it would open up to potential incoherences, like marking a method as coming from multiple sources.

File: analyzers/src/SonarAnalyzer.Common/Helpers/CodeCopiedFromAttribute.cs
Comment (antonioaversa) [inline analyzers/src/SonarAnalyzer.Common/Helpers/CodeCopiedFromAttribute.cs:41]: Any reason to have `Reasoning` a read-write property? 
I would make readonly, as all the other ones, and use a primary constructor:
```cs
[AttributeUsage(AttributeTargets.All, AllowMultiple = false, Inherited = false)]
public sealed class CodeCopiedFromAttribute(
    TrustedCodeSource TrustedCodeSource, string PermanentUrl, string Reasoning) : Attribute;
```

Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/CodeCopiedFromAttribute.cs:41]: Because it is optional.

Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/CodeCopiedFromAttribute.cs:41]: Default parameters for attributes are considered bad practise because of metadata bloat. It is only a single byte more, but still:
```
    .custom instance void MyAttribute::.ctor(string) = (
        01 00 ff 00 00
    )
    .custom instance void My2Attribute::.ctor() = (
        01 00 00 00
```
https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgDaBZAngCgJQC6qmWc+RKMAzAATw0DCNA3qje3bTACw0b4s2HAL6pRKVNTpw+WAILBgAJzAAjAK7AApjRA0FytZq2pWKDp1kGVG7TgDKhgHYBzGtoDOwGgF4aT9QgIPCF2MwtxcUkuGWw4ayNtXX1FG2NTVFDLGAQABhoAFS0vFhoXLWAAbhoPCurIoA=

File: analyzers/src/SonarAnalyzer.Common/Helpers/CodeCopiedFromAttribute.cs
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/CodeCopiedFromAttribute.cs:29]: It needs to be understood by the coverage tool but at the same time, we already copied enums to our code base and it might be useful to have that as a valid target. I don't think we should allow Assembly/Module though. Others, like fields/events, seem to be granular to me.

**Comment by martin-strecker-sonarsource** [comment]
@antonioaversa  and @zsolt-kolbay-sonarsource 

Ready for review.

@zsolt-kolbay-sonarsource This PR copies code from Roslyn, which will be used in the following PR https://github.com/SonarSource/sonar-dotnet/pull/8950

Please read this section https://docs.google.com/document/d/1WgxCWCFyKt2mI46U83fWkpea06NT-EmwWIa_XmvWLM0/edit#heading=h.rymzqwph8det on how we decided to copy code from Roslyn. Just so you know, I needed to roll back the new Attribute suggested there, as it was not working as intended. We decided to unblock the PR to use `[ExcludeFromCodeCoverage]` and code comments as an alternative to the Attribute and add an issue to investigate why the Attribute wasn't working (see https://github.com/SonarSource/sonar-dotnet/pull/8947#issuecomment-2056798852).

**Comment by martin-strecker-sonarsource** [comment]
Please also check the accepted issues:
https://sonarcloud.io/project/issues?resolutions=FALSE-POSITIVE%2CFIXED%2CWONTFIX&pullRequest=8947&id=sonaranalyzer-dotnet

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/ExpressionSyntaxExtensions.Roslyn.cs
Comment (antonioaversa) [inline analyzers/src/SonarAnalyzer.CSharp/Extensions/ExpressionSyntaxExtensions.Roslyn.cs:16]: In the VB.NET version below, I propose to add a documentation comment to explain what the method does. Maybe the best place to add such documentation would be the language or syntax facade, where the `IsWrittenTo` is wrapped.

File: analyzers/src/SonarAnalyzer.VisualBasic/Extensions/ExpressionSyntaxExtensions.Roslyn.cs
Comment (antonioaversa) [inline analyzers/src/SonarAnalyzer.VisualBasic/Extensions/ExpressionSyntaxExtensions.Roslyn.cs:14]: This method is excluded from coverage explicitely, at method level.
However, the methods below, such as `IsOnlyWrittenTo`, are not.
Is there a specific reason for that?

None of the methods in `Microsoft.CodeAnalysis.Shared.Extensions.SyntaxNodeExtensions` are directly marked as `[ExcludeFromCodeCoverage]`.

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.Roslyn.cs
Comment (antonioaversa) [inline analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.Roslyn.cs:15]: This (and also the methods below) is a bit different syntactically from the `Copy of` comments below.

I think it's fine, since we are going to introduce a dedicated attribute soon, that will make the format uniform.

An alternative would be to add a "normal" comment `Copy of` above `/// <summary>`. It would probably even be more correct since the documentation itself is a copy of the documentation in the Roslyn repository. 

Everything below such `// Copy of` comment would be 1-to-1 with the corresponding code on Roslyn, whereas this is not the case right now, where a diff would reveal added code in the middle of the documentation.

**Review by antonioaversa — CHANGES_REQUESTED** [review]

**Comment by antonioaversa** [comment]
@martin-strecker-sonarsource 
SonarCloud analysis gives 0 coverage due to the 2 lines uncovered in the C# and VB.NET facades:
```cs
    public override bool IsWrittenTo(SyntaxNode expression, SemanticModel semanticModel, CancellationToken cancellationToken) =>
        Cast<ExpressionSyntax>(expression).IsWrittenTo(semanticModel, cancellationToken);
```

While the `IsWrittenTo` extensions are copies from Roslyn, and as such are excluded from coverage, our wrappers in the facades are not, and I think they should be tested. 

Basic tests, just to make sure that the underlying extension method is called.
That should also give 100% coverage on non-Roslyn code.

What do you think?

File: analyzers/src/SonarAnalyzer.VisualBasic/Facade/VisualBasicSyntaxFacade.cs
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.VisualBasic/Facade/VisualBasicSyntaxFacade.cs:111]: The SyntaxFacade answers questions about a SyntaxTree in a language-agnostic way. Sometimes you need to query the semantic model before you can give a precise answer. The major benefit is having abstraction over the language. Any dependency on a semanticModel/compilation/cancelationToken/generated-code detector or any other service is not relevant here.

**Comment by martin-strecker-sonarsource** [comment]
> While the IsWrittenTo extensions are copies from Roslyn, and as such are excluded from coverage, our wrappers in the facades are not, and I think they should be tested.

Such tests are included in https://github.com/SonarSource/sonar-dotnet/pull/8950

File: analyzers/src/SonarAnalyzer.Common/Helpers/CodeCopiedFromAttribute.cs
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.Common/Helpers/CodeCopiedFromAttribute.cs]: Following the offline discussion we had, the attribute has been removed from this PR, since we could not make the coverage work correctly with it, and we want to prioritize the delivery of this PR, as other PRs depend on this.
@andrei-epure-sonarsource I am resolving this discussion.

File: analyzers/src/SonarAnalyzer.Common/Helpers/CodeCopiedFromAttribute.cs
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.Common/Helpers/CodeCopiedFromAttribute.cs]: I haven't though of that, thanks!

File: analyzers/src/SonarAnalyzer.Common/Helpers/CodeCopiedFromAttribute.cs
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.Common/Helpers/CodeCopiedFromAttribute.cs]: Alright, 1 byte doesn't seem enough to me to talk about "metadata bload", but fair enough.

**Comment by antonioaversa** [comment]
> I can not get coverage support for `CopiedCodeFrom` working. @antonioaversa @andrei-epure-sonarsource do you have any ideas what is missing in the coverage setup of the pipeline/SonarCloud? I generated the coverage file locally and it looks like so for methods annotated with the attribute. The important part is `skippedDueTo="Filter"`
> 
> ```
>         <Class>
>           <Summary numSequencePoints="0" visitedSequencePoints="0" numBranchPoints="0" visitedBranchPoints="0" sequenceCoverage="0" branchCoverage="0" maxCyclomaticComplexity="16" minCyclomaticComplexity="1" visitedClasses="0" numClasses="0" visitedMethods="0" numMethods="6" minCrapScore="0" maxCrapScore="0" />
>           <FullName>Microsoft.CodeAnalysis.VisualBasic.Extensions.ExpressionSyntaxExtensions</FullName>
>           <Methods>
>             <Method skippedDueTo="Filter" visited="false" cyclomaticComplexity="16" nPathComplexity="0" sequenceCoverage="0" branchCoverage="0" isConstructor="false" isStatic="true" isGetter="false" isSetter="false" crapScore="272">
>               <Summary numSequencePoints="0" visitedSequencePoints="0" numBranchPoints="0" visitedBranchPoints="0" sequenceCoverage="0" branchCoverage="0" maxCyclomaticComplexity="0" minCyclomaticComplexity="0" visitedClasses="0" numClasses="0" visitedMethods="0" numMethods="0" minCrapScore="272" maxCrapScore="272" />
>               <MetadataToken>100663316</MetadataToken>
>               <Name>System.Boolean Microsoft.CodeAnalysis.VisualBasic.Extensions.ExpressionSyntaxExtensions::IsWrittenTo(Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax,Microsoft.CodeAnalysis.SemanticModel,System.Threading.CancellationToken)</Name>
>               <SequencePoints />
>               <BranchPoints />
>               <MethodPoint vc="0" uspid="100663316" ordinal="0" offset="0" />
>             </Method>
>             
> ```
> 
> For a class it looks like so:
> 
> ```
>         <Class skippedDueTo="Filter">
>           <Summary numSequencePoints="0" visitedSequencePoints="0" numBranchPoints="0" visitedBranchPoints="0" sequenceCoverage="0" branchCoverage="0" maxCyclomaticComplexity="0" minCyclomaticComplexity="0" visitedClasses="0" numClasses="0" visitedMethods="0" numMethods="0" minCrapScore="0" maxCrapScore="0" />
>           <FullName>Microsoft.CodeAnalysis.CSharp.Extensions.ExpressionSyntaxExtensions</FullName>
>           <Methods />
>         </Class>
> ```
> 
> But in SonarCloud it is still reported as "uncovered by tests" in both cases.

Discussed offline, @martin-strecker-sonarsource is removing the attribute from this PR, as S6932 depends on this PR being delivered. We are going to reintroduce the attribute in a later PR.

@martin-strecker-sonarsource Do we have already an issue in the backlog to cover that gap? Can you please mention it in the description of this PR?

File: analyzers/src/SonarAnalyzer.VisualBasic/Facade/VisualBasicSyntaxFacade.cs
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.VisualBasic/Facade/VisualBasicSyntaxFacade.cs:112]: Thanks for explaining your point of view. I will adopt in the future the same approach, when in doubt about whether to add a method in `SyntaxFacade` or `ILanguageFacade`.

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/ExpressionSyntaxExtensions.Roslyn.cs
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/Extensions/ExpressionSyntaxExtensions.Roslyn.cs:22]: I see you went for inheriting the doc from the copied extension method to the facade. Good solution as well.
From the perspective of the user of the method in the facade, it shouldn't change much (provided he is in the context of the IDE, otherwise he has to follow the link in the doc).

**Comment by antonioaversa** [comment]
> > While the IsWrittenTo extensions are copies from Roslyn, and as such are excluded from coverage, our wrappers in the facades are not, and I think they should be tested.
> 
> Such tests are included in #8950

Alright, that means however that this PR will need to be merged by a champion.
@costin-zaharia-sonarsource

**Review by antonioaversa — APPROVED** [review]
LGTM! Let's merge this one and unblock the following PRs.

@zsolt-kolbay-sonarsource I would suggest to merge and not wait for a second review, given how late we are with this and how much depend on this PR. The quality of the PR is really good, the approach has been extensively discussed in the "How We Work" circle.

@martin-strecker-sonarsource Please, check [this comment](https://github.com/SonarSource/sonar-dotnet/pull/8947#issuecomment-2074235179)

@costin-zaharia-sonarsource Please check [this comment](https://github.com/SonarSource/sonar-dotnet/pull/8947#issuecomment-2074256629)

**Comment by martin-strecker-sonarsource** [comment]
> Do we have already an issue in the backlog to cover that gap? Can you please mention it in the description of this PR?

https://github.com/SonarSource/sonar-dotnet/issues/9201

**Pull request merged**
